File: lecture_09/primes.py
"""
A simple python program to test CPU performance by
generating a series of prime numbers
"""
from numba import njit, boolean, uint64, void

NUM_PRIMES = 100_000


@njit(boolean(uint64))
#@jit(boolean(uint64), nopython=True )
def is_prime(number):
    "Returns True if 'number' is prime"
    limit = int(number ** 0.5) + 1
    for d in range(2, limit):
        if number % d == 0:
            return False
    return True


@njit(void(uint64))
def gen_primes(num_primes):
    """Prints the first 'num_primes' series of prime numbers"""
    v = 2
    while num_primes:
        if is_prime(v):
            print(v)
            num_primes -= 1
        v += 1


if __name__ == '__main__':
    from time import time
    import sys

    print(f"Print the first {NUM_PRIMES} prime numbers...", file=sys.stderr)
    start_time = time()
    gen_primes(NUM_PRIMES)
    duration = time() - start_time
    print(f"print_primes({NUM_PRIMES}) took {duration:.6f} seconds.",
          file=sys.stderr)

File: lecture_04/primes.rs
use std::time::Instant;
const NUM_PRIMES: u64 = 100_000;

fn is_prime(number: u64) -> bool {
    let limit = (number as f64).sqrt() as u64 + 1;
    for d in 2..limit {
        if number % d == 0 {
            return false;
        }
    }
    return true;
}

fn gen_primes(mut num_primes: u64) {
    let mut v = 2;
    while num_primes > 0 {
        if is_prime(v) {
            println!("{}", v);
            num_primes -= 1;
        }
        v += 1
    }
}

fn main() {
    eprintln!("Generating {} prime numbers...", NUM_PRIMES);
    let start_time = Instant::now();
    gen_primes(NUM_PRIMES);
    let duration = start_time.elapsed();
    eprintln!("gen_primes({}) took {:.6} seconds.",
            NUM_PRIMES,
            duration.as_secs_f64()
    );
}
